Cgroup Resource sudo integration test

sharanyad opened the pull request:
### Summary
Integration test for task's cgroup resource creation and deletion as sudo 

### Implementation details
- added new make target `run-sudo-tests` for running integration tests as sudo
- added tags `!sudo` for other test files since `-tags sudo` runs all tests without sudo build tags too. 
- separated out some common functions needed for integration tests into `common_integ_test.go` and `common_unix_integ_test.go`.

### Testing
- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

petderek commented on agent/app/agent_integ_test.go:
```diff
@@ -1,4 +1,4 @@
-// +build integration
+// +build !sudo,integration
```
ex: if sudo implies integration, this file doesn't need to be changed.

petderek commented on Makefile:
```diff
@@ -220,9 +221,12 @@ cni-plugins: get-cni-sources .out-stamp
 	. ./scripts/shared_env && go test -race -tags integration -timeout=7m -v ./agent/engine/... ./agent/stats/... ./agent/app/...
 
+run-sudo-integ-tests:
+	. ./scripts/shared_env && sudo -E ${GO_EXECUTABLE} test -race -tags 'sudo integration' -timeout=5m -v ./agent/engine/... ./agent/stats/... ./agent/app/...
```
I think it simplifies the build logic if you only use a 'sudo' built target that implies integration. You wouldn't need to add the !sudo to every additional integ test.

sharanyad replied:
My concern about using just `sudo` was there may be different types of tests in the future that would need to run as `sudo`. Hence kept it to two tags. What do you think?

petderek replied:
I had considered that as well, but if we are interacting with host/docker resources its basically an integ test by definition.

petderek commented on agent/engine/engine_sudo_linux_integ_test.go:
```diff
@@ -0,0 +1,80 @@
+// +build sudo,linux,integration
```
ex: this file would just have 'sudo,linux' as build tags

sharanyad commented on Makefile:
```diff
@@ -220,9 +221,12 @@ cni-plugins: get-cni-sources .out-stamp
 	. ./scripts/shared_env && go test -race -tags integration -timeout=7m -v ./agent/engine/... ./agent/stats/... ./agent/app/...
 
+run-sudo-integ-tests:
+	. ./scripts/shared_env && sudo -E ${GO_EXECUTABLE} test -race -tags 'sudo integration' -timeout=5m -v ./agent/engine/... ./agent/stats/... ./agent/app/...
```
`-tags` only accepts space separated lists.

petderek replied:
I had thought this same thing! The syntax is weirder than I expected.

sharanyad commented on agent/app/agent_capability_test.go:
```diff
@@ -1,3 +1,4 @@
+// +build !integration
```
I had to add `!sudo` and `!integration` since `-tags 'sudo integration'` would run tests in files without the tags `sudo` or `integration` too, which is weird.

petderek commented on Makefile:
```diff
@@ -220,9 +221,12 @@ cni-plugins: get-cni-sources .out-stamp
 	. ./scripts/shared_env && go test -race -tags integration -timeout=7m -v ./agent/engine/... ./agent/stats/... ./agent/app/...
 
+run-sudo-integ-tests:
+	. ./scripts/shared_env && sudo -E ${GO_EXECUTABLE} test -race -tags sudo -timeout=5m -v ./agent/engine/... ./agent/stats/... ./agent/app/...
```
the only sudo test is in `./agent/engine`

You could get rid of the stats and app calls here, and then get rid of the !sudo tags in app and stats.

sharanyad replied:
done

sharanyad commented:
> it will became more complicated if we are adding new flags in the future. So instead of using the negative tag !, can we just add tags to every test file? 

We would still need negative tags for the newly introduced tags. If the new tag is not present in a file, then go build assumes that it is present by default, which is why we would need the `!` flag. Does that make sense?

petderek commented:
> We don't run unit tests via integration tag today right?

We explicitly avoid that by using the `!integration` tag. `go test` would try to run them otherwise. If we use `unit`, we no longer need the `!integration` tag because its implied.

sharanyad commented:
Update: new changes include adding 'unit' build tag for unit test files, to avoid adding `!` tags when new ones are introduced. 
Changes: added 'unit' build tag, updated copyright year in the headers.

petderek commented on Makefile:
```diff
@@ -220,9 +221,12 @@ cni-plugins: get-cni-sources .out-stamp
 	@echo "Built amazon-ecs-cni-plugins successfully."
 
-run-integ-tests: test-registry gremlin container-health-check-image
+run-integ-tests: test-registry gremlin container-health-check-image run-sudo-tests
```
I'd prefer to keep the sudo business separated

sharanyad replied:
Since sudo tests are nothing but integ tests, I opted to keep it as part of `run-integ-tests`.

petderek commented on agent/acs/client/acs_client_test.go:
```diff
@@ -1,4 +1,5 @@
-// Copyright 2014-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// +build unit
```
> To distinguish build constraints from package documentation, a series of build constraints must be followed by a blank line.

I'm not sure if this is enforced? But we might want to change it.

https://golang.org/pkg/go/build/#hdr-Build_Constraints

sharanyad replied:
Looks like this is not enforced. But better to change this.
